docs/source/starthere/fundamentals.rst

```diff
@@ -0,0 +1,240 @@
+
+.. list-table:: Simplified examples of a model and config.
+    :widths: 1 1
+    :header-rows: 0
+
+    * - .. code-block:: python
+	  :caption: NeMo model definition (Python pseudocode)
+	  :linenos:
+	  :emphasize-lines: 4, 7, 10, 13, 16, 20
+
+	  class ExampleEncDecModel:
+	      # cfg is passed so it only contains "model" section
+	      def __init__(self, cfg, trainer):
+	          self.tokenizer = init_from_cfg(cfg.tokenizer)
+
+
+	          self.encoder = init_from_cfg(cfg.encoder)
+
+
+	          self.decoder = init_from_cfg(cfg.decoder)
+
+
+	          self.loss = init_from_cfg(cfg.loss)
+
+
+		  # optimizer configured via parent class
+
+
+	      def setup_training_data(self, cfg):
+	          self.train_dl = init_dl_from_cfg(cfg.train_ds)
+
+	      def forward(self, batch):
+	          # forward pass defined,
+		  # as is standard for PyTorch models
+	          ...
+
+	      def training_step(self, batch):
+	          log_probs = self.forward(batch)
+	          loss = self.loss(log_probs, labels)
+	          return loss
+
+
+      - .. code-block:: yaml
+	  :caption: Experiment config (YAML)
+	  :linenos:
+	  :emphasize-lines: 4, 7, 10, 13, 16, 20
+
+	  #
+	  # desired configuration of the NeMo model
+	  model:
+	      tokenizer:
+	       ...
+
+	      encoder:
+	       ...
+
+	      decoder:
+	       ...
+
+	      loss:
+	       ...
+
+	      optim:
+	       ...
+
+
+	      train_ds:
+	       ...
+
+	  # desired configuration of the
+	  # PyTorch Lightning trainer object
+	  trainer:
+	      ...
+
+
+Configuring and training NeMo models
+------------------------------------
+
+During initialization of the model, key parameters are read from the config (``cfg``), which gets passed in to the model construtor (left panel above, line 2).
+
+The other object passed into the model's constructor is a PyTorch Lightning ``trainer`` object, which handles the training process. The trainer will take care of the standard training `boilerplate <https://lightning.ai/docs/pytorch/stable/common/trainer.html#under-the-hood>`__. For things that are not standard, PTL will refer to any specific methods that we may have defined in our NeMo model. For example, PTL requires every model to have a specified ``training_step`` method (left panel above, line 15).
+
+The configuration of the trainer is also specified in the config (right panel above, line 20 onwards). This will include parameters such as (number of) ``devices``, ``max_steps``, (numerical) ``precision`` and `more <https://lightning.ai/docs/pytorch/stable/common/trainer.html#trainer-class-api>`__.
+
+
+Example training script
+-----------------------
+
+Putting it all together, here is an example training script for our ``ExampleEncDecModel`` model. We highlight the 3 most important lines, which put together everything we discussed in the previous section.
```

Comment:
suggested revision.

Below is an example training script for our ``ExampleEncDecModel`` model. We highlight the three most important lines that combine everything we discussed in the previous section: